### PR TITLE
V1-177 - used debounce and fixed issue with many requests

### PR DESCRIPTION
**Pull request by zhuk-vezunchikz**

### CODE REVIEW

File: frontend/src/components/Layout/Header/SearchCourses/SearchResultItem.tsx [hunk 1]
```diff
@@ -14,23 +14,23 @@ import { Image, CourseTitle, SearchResultCourse } from './styled';
...
 }) => (
   <Link
     key={course._id}
     to={
-      foundInMyCourses
-        ? `${PATHS.myCourses}/${foundInMyCourses}`
+      foundInMyCoursesId
+        ? `${PATHS.myCourses}/${foundInMyCoursesId}`
         : `${PATHS.coursesList}/${course._id}`
     }
```
Comment (k-krasilnikova) [inline frontend/src/components/Layout/Header/SearchCourses/SearchResultItem.tsx:35]: Let's wait for Daniel PR to be merged and rebase on develop: those lines will be changed

File: frontend/src/components/Layout/Header/SearchCourses/SearchResult.tsx [hunk 2]
```diff
@@ -11,19 +13,43 @@ interface CoursesFound {
...
+  };
+
+  useEffect(() => {
+    const getClientCourses = async () => {
+      const clientCourses = await searchClientCourses();
+      coursesFound.forEach((course) => findCourse(clientCourses, course.title));
+    };
+
+    getClientCourses();
+    // eslint-disable-next-line react-hooks/exhaustive-deps
```
Comment (k-krasilnikova) [inline frontend/src/components/Layout/Header/SearchCourses/SearchResult.tsx:35]: Are u sure that we need to use async/await in our component? Where is hooks?

File: frontend/src/components/Layout/Header/SearchCourses/SearchResult.tsx [hunk 3]
```diff
@@ -11,19 +13,43 @@ interface CoursesFound {
...
-    ) : (
-      <NoSearchResults>No results</NoSearchResults>
-    )}
-  </SearchResultWrapper>
-);
+const SearchResult: React.FC<CoursesFound> = ({ coursesFound }) => {
+  const [foundInMyCourses, setFoundInMyCourses] = useState<string | undefined>(undefined);
+
+  const findCourse = (courses: Array<ClientCourse>, title: string): void => {
+    for (let clientCourse = 0; clientCourse < courses.length; clientCourse += 1) {
```
Comment (k-krasilnikova) [inline frontend/src/components/Layout/Header/SearchCourses/SearchResult.tsx:20]: `clientCourse` seems like an object, but not number. Rename this variable please

File: frontend/src/components/Layout/Header/SearchCourses/SearchCoursesContainer.tsx [hunk 4]
```diff
@@ -14,10 +15,12 @@ const SearchCoursesContainer: React.FC = () => {
 
   const { enqueueSnackbar } = useSnackbar();
 
+  const debouncedSearchValue = useDebounce(searchInputValue, 2000);
```
Comment (k-krasilnikova) [inline frontend/src/components/Layout/Header/SearchCourses/SearchCoursesContainer.tsx:18]: Move 2000 to the constants please

**Review by k-krasilnikova — CHANGES_REQUESTED** [review]
Good job, but some changes required

File: frontend/src/api/myCourses/searchClientCourses.ts [hunk 5]
```diff
@@ -1,17 +1,17 @@
...
+
 import { apiClientWrapper } from 'api/base';
 import { API } from 'constants/routes';
-import { REQUEST_ERRORS } from 'constants/authConstants';
 import { ClientCourse } from 'types/clientCourse';
 
-const searchClientCourses = async (): Promise<ClientCourse[]> => {
-  const apiClient = apiClientWrapper();
-  try {
+const useSearchClientCourses = (): UseQueryResult<Array<ClientCourse>, AxiosError> => {
```
[the remark this answers is not in the document]
Comment (zhuk-vezunchikz) [reply frontend/src/api/myCourses/searchClientCourses.ts:8]: done

File: frontend/src/components/Layout/Header/SearchCourses/SearchCoursesContainer.tsx [hunk 6]
```diff
@@ -27,10 +31,10 @@ const SearchCoursesContainer: React.FC = () => {
           }
         }
       };
-      getCourses(searchInputValue);
-    }, 1000);
-    return () => clearTimeout(timer);
-  }, [searchInputValue, enqueueSnackbar]);
+      getCourses(debouncedSearchValue);
+    }
+    // eslint-disable-next-line react-hooks/exhaustive-deps
```
[the remark this answers is not in the document]
Comment (zhuk-vezunchikz) [reply frontend/src/components/Layout/Header/SearchCourses/SearchCoursesContainer.tsx]: done

File: frontend/src/components/Layout/Header/SearchCourses/SearchCoursesContainer.tsx [hunk 7]
```diff
@@ -14,10 +16,12 @@ const SearchCoursesContainer: React.FC = () => {
...
   const { enqueueSnackbar } = useSnackbar();
 
+  const debouncedSearchValue = useDebounce(searchInputValue, SEARCH_DEBOUNCE_TIME);
+
   useEffect(() => {
-    const timer = setTimeout(() => {
+    if (debouncedSearchValue) {
       const getCourses = async (search: string) => {
-        if (searchInputValue.length) {
+        if (debouncedSearchValue.length) {
```
[the remark this answers is not in the document]
Comment (zhuk-vezunchikz) [reply frontend/src/components/Layout/Header/SearchCourses/SearchCoursesContainer.tsx]: done

File: frontend/src/components/Layout/Header/SearchCourses/SearchCoursesContainer.tsx (hunk 4 above)
Comment (zhuk-vezunchikz) [reply frontend/src/components/Layout/Header/SearchCourses/SearchCoursesContainer.tsx]: done

File: frontend/src/components/Layout/Header/SearchCourses/SearchResult.tsx (hunk 3 above)
Comment (zhuk-vezunchikz) [reply frontend/src/components/Layout/Header/SearchCourses/SearchResult.tsx]: done

File: frontend/src/components/Layout/Header/SearchCourses/SearchResult.tsx (hunk 2 above)
Comment (zhuk-vezunchikz) [reply frontend/src/components/Layout/Header/SearchCourses/SearchResult.tsx]: done

File: frontend/src/components/Layout/Header/SearchCourses/SearchResultItem.tsx (hunk 1 above)
Comment (zhuk-vezunchikz) [reply frontend/src/components/Layout/Header/SearchCourses/SearchResultItem.tsx]: done

File: frontend/src/api/courses/searchAllCourses.ts [hunk 8]
```diff
@@ -1,22 +1,18 @@
...
-import { REQUEST_ERRORS } from 'constants/authConstants';
 import { Course } from 'types/course';
 
-const searchAllCourses = async (
-  courseName: string,
-): Promise<AxiosResponse<Course[], AxiosError>> => {
-  const apiClient = apiClientWrapper();
-  try {
+const useSearchAllCourses = (courseName: string): UseQueryResult<Array<Course>, AxiosError> => {
+  return useQuery(['courses', courseName], async () => {
```
Comment (irina-sinitsyna) [inline frontend/src/api/courses/searchAllCourses.ts:9]: please replace string with QUERY_KEY constant

File: frontend/src/api/myCourses/searchClientCourses.ts [hunk 9]
```diff
@@ -1,17 +1,17 @@
...
+  return useQuery('search', async () => {
```
Comment (irina-sinitsyna) [inline frontend/src/api/myCourses/searchClientCourses.ts:9]: QUERY_KEY

File: frontend/src/components/Layout/Header/SearchCourses/SearchResult.tsx [hunk 10]
```diff
@@ -11,19 +13,41 @@ interface CoursesFound {
...
+      {coursesFound.length ? (
+        coursesFound.map((course, id, array) => (
+          <SearchResultItemContainer
+            course={course}
+            addDivider={id < array.length + LAST_ARRAY_ITEM}
+            foundInMyCoursesId={foundInMyCourses}
+          />
+        ))
+      ) : (
+        <NoSearchResults>No results</NoSearchResults>
```
Comment (irina-sinitsyna) [inline frontend/src/components/Layout/Header/SearchCourses/SearchResult.tsx:47]: please add this string to constants/messages and reuse here

**Review by irina-sinitsyna — CHANGES_REQUESTED** [review]

File: frontend/src/api/courses/searchAllCourses.ts (hunk 8 above)
Comment (zhuk-vezunchikz) [reply frontend/src/api/courses/searchAllCourses.ts]: done

File: frontend/src/api/myCourses/searchClientCourses.ts (hunk 9 above)
Comment (zhuk-vezunchikz) [reply frontend/src/api/myCourses/searchClientCourses.ts]: done

**Pull request merged**